src/packet/bwe/mod.rs

```diff
@@ -231,3 +240,149 @@ impl fmt::Display for BandwidthUsage {
         }
     }
 }
+
+/// Sliding window [`SeqNo`]s storage.
+///
+/// Only remembers the last [`SIZE`] packets added.
+#[derive(Debug)]
+struct HandledPacketsTracker<const SIZE: usize> {
+    /// Recently added packets.
+    history: [Option<u16>; SIZE],
+
+    /// Queue that tracks added packets order so older packets are removed.
+    queue: VecDeque<u16>,
+}
+
+impl<const SIZE: usize> HandledPacketsTracker<SIZE> {
+    /// Remembers the give [`SeqNo`].
+    ///
+    /// Expects somewhat sequential data with reordering no more than configured
+    /// [`SIZE`].
+    fn add(&mut self, seq: SeqNo) {
+        let seq = seq.as_u16();
+        let history_idx = seq as usize % SIZE;
+
+        self.queue.push_back(seq);
+        if self.queue.len() == SIZE {
+            let to_remove = self.queue.pop_front().unwrap();
+            let remove_idx = to_remove as usize % SIZE;
+            if self.history[remove_idx] == Some(seq) {
```

Comment:
Isn't the invariant here that we expect `self.history[x % SIZE] == Some(x)` if `self.queue` contains `x`? If so shouldn't we always zero out the history here?

Reply:
This is to account for possible collisions.